lib/ast/AST.cpp

```diff
@@ -928,6 +928,106 @@ sptr<KOREPattern> KORECompositePattern::filterSubstitution(PrettyPrintData const
   }
 }
 
+sptr<KOREPattern> KORECompositePattern::expandMacros(SubsortMap const& subsorts, SymbolMap const& overloads, std::vector<ptr<KOREDeclaration>> const& macros, bool reverse, std::set<size_t> &appliedRules) {
+  size_t i = 0;
+  for (auto &decl : macros) {
+    if ((decl->getAttributes().count("macro") || decl->getAttributes().count("macro-rec")) && reverse) {
+      i++;
+      continue;
+    }
+    auto axiom = dynamic_cast<KOREAxiomDeclaration *>(decl.get());
+    auto equals = dynamic_cast<KORECompositePattern *>(axiom->getPattern().get());
+    auto lhs = equals->arguments[reverse ? 1 : 0];
+    auto rhs = equals->arguments[reverse ? 0 : 1];
+    substitution subst;
+    bool matches = lhs->matches(subst, subsorts, overloads, shared_from_this());
+    if (matches && (decl->getAttributes().count("macro-rec") || decl->getAttributes().count("alias-rec") || !appliedRules.count(i))) {
+      std::set<size_t> oldAppliedRules = appliedRules;
+      appliedRules.insert(i);
+      auto result = rhs->substitute(subst)->expandMacros(subsorts, overloads, macros, reverse, appliedRules);
+      appliedRules = oldAppliedRules;
```

Comment:
appliedRules is a reference so yes, this is necessary.

Reply:
Ok, I actually meant this: Is the referred value used anywhere after the `rhs->substitute(subst)->expandMacros` call? As I see it:
* KorePattern::expandMacros with 4 arguments passes a reference to a local object when it adds `appliedRules` as an extra argument. It never uses it after the call.
* `KORECompositePattern::expandMacros` with 5 arguments uses it only inside an `if` statement that calls `expandMacros` with a changed `appliedRules`, then returns from the function, so it does not use this value after the call.
* I think that the above are the only calls where `appliedRules` is used as an argument, so I would say that there is no way of using it after the call.

Assuming that's correct, then, if you still want to restore the old value, then how about this: you could make `appliedRules` be a const reference, then you could rewrite the `if` like this:
```c
if (stuff) {
￼      std::set<size_t> newAppliedRules = appliedRules;
￼      newAppliedRules.insert(i);
￼      return rhs->substitute(subst)->expandMacros(subsorts, overloads, macros, reverse, newAppliedRules);
}
```

Reply:
I can tell you that it IS used after the call because it's used after the function returns. I can't make it a const reference because it's actually modified. Regardless, this code is simply an accurate translation of some Java code that this entire algorithm is based on, and I'm not going to change it to be less like the original just because it is allegedly stylistically better, because it's not at all clear to me that the change you are proposing preserves behavior. And even if it in fact does, it's better that the code resemble the original as much as possible.